core/include/traccc/fitting/triplet_fit/triplet_fitter.hpp

```diff
@@ -0,0 +1,931 @@
+        : m_detector(det), m_field(field), m_cfg(cfg) {}
+
+    // Triplet struct
+    struct triplet {
+
+        /// Default construct
+        triplet() = default;
+
+        /// Construct with three hits
+        ///
+        /// @param hit_0 position of first hit
+        /// @param hit_1 position of second hit
+        /// @param hit_2 position of third hit
+        triplet(const point3& hit_0, const point3& hit_1, const point3& hit_2)
+            : m_hit_0(hit_0), m_hit_1(hit_1), m_hit_2(hit_2) {}
+
+        // global positions of three hits
+        point3 m_hit_0;
+        point3 m_hit_1;
+        point3 m_hit_2;
+
+        // triplet parameters
+        scalar m_phi_0{};
+        scalar m_theta_0{};
+        scalar m_rho_phi{};
+        scalar m_rho_theta{};
+
+        // hit position derivatives
+        vecmem::vector<scalar> m_h_thet;
+        vecmem::vector<scalar> m_h_phi;
+
+        // Measurements for getting variances (hit shifts)
+        // and surface orientation (scattering estimation)
+        std::array<measurement, 3u> m_meas;
+
+        // Estimated values
+        scalar m_sigma_MS{};  // MS uncertainty
+        scalar m_theta{};     // Polar angle
+    };
+
+    /// Helper function - Initialize fitter
+    ///
+    /// @param in_track_states input track states (from measurements)
+    ///
+    TRACCC_HOST_DEVICE
+    void init_fitter(
+        const vecmem::vector<track_state<algebra_type>>& in_track_states) {
+
+        // m_track_states(vecmem::device_vector<track_state<algebra_type>>(in_track_states));
+
+        m_track_states = in_track_states;
+
+        // Clear triplets from the last track
+        m_triplets.clear();
+    }
+
+    /// Helper function - Make triplets
+    ///
+    /// Makes triplets from consecutive measurements on the track
+    ///
+    TRACCC_HOST_DEVICE
+    void make_triplets() {
+
+        // std::cout << "Making triplets\n";
+
+        size_t n_triplets = m_track_states.size() - 2;
+
+        m_triplets.reserve(n_triplets);
+
+        // loop over measurements (track states) in candidate
+        for (size_t i = 0; i < n_triplets; ++i) {
+
+            // Get track states (and measurements)
+            const track_state<algebra_type>& state_0 = m_track_states[i];
+            const track_state<algebra_type>& state_1 = m_track_states[i + 1];
+            const track_state<algebra_type>& state_2 = m_track_states[i + 2];
+
+            const measurement& meas_0 = state_0.get_measurement();
+            const measurement& meas_1 = state_1.get_measurement();
+            const measurement& meas_2 = state_2.get_measurement();
+
+            // Get surfaces
+            detray::tracking_surface meas_0_sf(m_detector, meas_0.surface_link);
+            detray::tracking_surface meas_1_sf(m_detector, meas_1.surface_link);
+            detray::tracking_surface meas_2_sf(m_detector, meas_2.surface_link);
+
+            point2 loc_2d_0{meas_0.local[0], meas_0.local[1]};
```

Comment:
Do you guarantee somewhere that all three measurements are from pixels? This will not work on 1D measurements (strip, drift chamber, straw tube).

Reply:
I tested this for also strip measurements, and it worked fine. From my experience I saw always two dimensions. Also the `local` member variable in the structure is two dimensional

Reply:
Code may run w/o error but a "single" strip measurement is not supposed to be used for `bound_to_global`. It is just going to produce wrong result

It is correct that measurement has two dimensional local member variable but one of the two elements is not used for strip measurement - that's why there is a `meas_dim` member variable.